packages/camelot/src/Camelot.ts

```diff
@@ -6,105 +6,147 @@ import {
 import { type Address } from 'viem'
 import { CHAIN_ID_ARRAY, ARBITRUM_CHAIN_ID } from './chain-ids'
 import { DEFAULT_TOKEN_LIST } from './contract-addresses'
-import { buildPathQuery, Tokens } from './utils'
-import { CAMELOT_ABI, PARASWAP_ABI } from './abi'
+import { buildV2PathQuery, buildV3PathQuery, Tokens } from './utils'
 import {
-  CAMELOT_ROUTER,
+  CAMELOT_V2_ABI,
+  CAMELOT_V3_EXACT_INPUT_ABI,
+  CAMELOT_V3_EXACT_OUTPUT_ABI,
+  PARASWAP_ABI,
+} from './abi'
+import {
+  CAMELOT_V2_ROUTER,
+  CAMELOT_V3_ROUTER,
   PARASWAP_ROUTER,
-  ETH_ADDRESS,
+  INTERNAL_ETH_ADDRESS,
 } from './contract-addresses'
 
-const isValidContractAddress = (address: Address) => {
-  return (
-    address?.toLowerCase() === CAMELOT_ROUTER.toLowerCase() ||
-    address?.toLowerCase() === PARASWAP_ROUTER.toLowerCase()
-  )
-}
-
 export const swap = async (
   swap: SwapActionParams,
 ): Promise<TransactionFilter> => {
-  const {
-    chainId,
-    contractAddress,
-    tokenIn,
-    tokenOut,
-    amountIn,
-    amountOut,
-    recipient,
-  } = swap
+  const { chainId, tokenIn, tokenOut, amountIn, amountOut, recipient } = swap
 
   const ethUsedIn = tokenIn === Tokens.ETH
   const ethUsedOut = tokenOut === Tokens.ETH
 
-  if (contractAddress && !isValidContractAddress(contractAddress)) {
-    throw new Error('Invalid Contract Address')
-  }
+  const tokenInOrEth = ethUsedIn ? INTERNAL_ETH_ADDRESS : tokenIn
+  const tokenOutOrEth = ethUsedOut ? INTERNAL_ETH_ADDRESS : tokenOut
+  const tokenInOrWeth = ethUsedIn ? Tokens.WETH : tokenIn
+  const tokenOutOrWeth = ethUsedOut ? Tokens.WETH : tokenOut
 
   return compressJson({
     chainId: chainId,
-    to: { $or: [CAMELOT_ROUTER.toLowerCase(), PARASWAP_ROUTER.toLowerCase()] },
+    from: recipient,
+    to: {
+      $or: [
+        CAMELOT_V2_ROUTER.toLowerCase(),
+        CAMELOT_V3_ROUTER.toLowerCase(),
+        PARASWAP_ROUTER.toLowerCase(),
+      ],
+    },
     value: ethUsedIn ? amountIn : undefined,
     input: {
-      $abi: [...CAMELOT_ABI, ...PARASWAP_ABI],
       $or: [
         {
           // camelotV2 swap
-          to: recipient,
-          path: buildPathQuery(ethUsedIn ? Tokens.WETH : tokenIn, tokenOut),
```

Comment:
This was an issue with the current plugin implementation. if tokenOut is selected as ETH, it needs to be converted to WETH in the path array.

Reply:
Thanks for the callout